### ISSUE TITLE
Compatibility issues in Fall Creators Update - probably ValueTuple-related

### ISSUE THREAD

**Issue by mkosieradzki** [Microsoft/dotnet-framework-early-access#9]
# Version info 
Microsoft Windows Version 1709 (OS Build 16299.19)

# General
First of all please forgive me I am not attaching full repro. I have lost entire day after upgrading to Fall Creators Update to track down this issue and I am way too tired to create at the moment proper reproduction, however I creating this issue now (even without full repro), because it might somehow affect rollout of Fall Creators Update.

My scenario:
Lib1 - library targeting .NET Standard 2.0
It has an interface that has a member that returns a ValueTuple (more precisely: `Task<(T1 F1, T2 F2)>`).

App1 - Service Fabric application (I have not tried to isolate this to a different kind of application yet) targeting .NET Framework 4.6.1
Inside this app I implement aforementioned interface from Lib1).

This application compiles and runs correctly on Windows 10 Creators Update.
This application fails on Windows 10 Fall Creators Update with TypeLoadException `Method 'XXX' in type 'YYY' from assembly 'ZZZ' does not have an implementation.`

I have lost a lot of time because I was also updating Service Fabric today and was sure it was rather related to some issues in new project system etc. However at last I have downgraded Service Fabric to previous version and the issue kept occurring. On my second machine with Windows 10 Creators Update everything works as expected.

I have noticed somewhere that Fall Creators Update is shipped with .NET 4.7.1 which has a potentially breaking change in ValueTuple implementation.

Finally I decided to try to change the .NET framework version in App1 to .NET 4.7 and everything started to work correctly.

**Comment by mkosieradzki** [comment]
[ConsoleApp2.zip](https://github.com/Microsoft/dotnet-framework-early-access/files/1385063/ConsoleApp2.zip)

OK got repro -> apparently I have just created a project according to description above and got full repro.

Try to run ConsoleApp2 (but only on a computer with Fall Creators Update or (my guess with .NET 4.7.1 installed) - you will get:
```
An unhandled exception of type 'System.TypeLoadException' occurred in Unknown Module.
Method 'Method' in type 'ConsoleApp2.MyClass' from assembly 'ConsoleApp2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.
```

After switching target framework to net47 in ConsoleApp2 app starts to work.

**Comment by AlexGhiondea** [comment]
Thanks @mkosieradzki for your report! I was able to repro the issue.

Would you mind trying to remove the binding redirect for System.ValueTuple from your .exe.config file?

**Comment by mkosieradzki** [comment]
@AlexGhiondea in which thread should we discuss?

I have not created explicitly any binding redirects. If you take a close look at the attached zip file it is ultra-minimalistic. Probably if I remove manually the binding redirect from my app it should work according to what you have written, should I try to do this? Would such an experiment have any value for you?

**Comment by AlexGhiondea** [comment]
Yes, please try to remove the ones that are autogenerated.

**Comment by mkosieradzki** [comment]
I am sorry I have responded in the wrong thread. After manually removing this binding redirect everything works as expected.

**Comment by jcouv** [comment]
@AlexGhiondea @weshaggard Thanks for the investigation. Does this only affect ValueTuple, or other types too? 
Is there a chance we can find a better solution than manually editing auto-generated redirects? Could we fix how they are generated?

**Comment by mkosieradzki** [comment]
How are you planning to fix binding redirects for already released applications? Both solutions are not a real solutions.

.NET Framework 4.7.1. should have not been released with such bugs, and you knew about this issue before release.

**Comment by jnm2** [comment]
If you remove the binding redirects, don't you also lose the ability to run your net461 application on net461–net47?

**Comment by IanYates** [comment]
@jnm2 - I had the same thoughts.  Curious to see if there's a better workaround, particularly now as Win 10 1709 is being rolled out with net471 included.

**Comment by kevingosse** [comment]
I'm quite a bit worried about this issue, as we're precisely about to migrate our libraries to netstandard 2.0 with a .NET 4.6.1 runtime. As of today this wouldn't cause problems as we are in a closed environment and all our servers are running .NET 4.6.1. But I'm afraid that by doing so, we may be locking ourselves out and transform a future migration to a more recent version of the .net framework into a nightmare.

What is Microsoft's stance on the issue? Can we expect that the compatibility issue will be fixed in a future update (either a servicing update for 4.7.1, or a 4.7.2), or do you consider that "what is done is done" and we need to take that into account for our migration plans?

**Comment by jcouv** [comment]
@kevingosse We're working on resolving this problem in 4.7.2. We're still reviewing a proposed design/solution.

**Comment by AlexGhiondea** [comment]
@kevingosse we are also looking into ways that we can deliver a fix to this for .NET 4.7.1 in the next few months.

**Comment by bording** [comment]
The addition of the netstandard assembly to 4.7.1 has impacted more scenarios than just the TypeLoadExceptions mentioned here so far.

Because the assembly is now in the GAC, it's being loaded from the GAC even when targeting < 4.7.1. This means that various assembly loading scenarios can start failing as well. For example, see https://github.com/Particular/NServiceBus/issues/5047.

@AlexGhiondea @jcouv Are the solutions you're looking at capable of addressing this problem as well?

Overall, these compatibility problems feel like something critical enough that they need to be fixed sooner rather than later. I do hope that "in the next few months" isn't really how long it's going to be before we see this addressed!

**Comment by karelz** [comment]
@bording the only repro and symptoms we're aware of are described here and in https://github.com/dotnet/corefx/issues/24653. We documented it as [known 4.7.1 issue](https://github.com/Microsoft/dotnet/blob/master/releases/net471/KnownIssues/510901-BCL%20Apps%20targeting%20.NET-4.6.1%20that%20use%20.NET%20Standard%20libraries%20might%20be%20broken%20when%20running%20on%20.NET%204.7.1.md)

If you think there are other problems, please give us repro steps. The issue you linked is quite general (not related to ValueTuple and similar types). We're not sure how exactly to repro it and root cause it. It may be the same problem, or it may be another problem. Thanks!

**Comment by bording** [comment]
@karelz I'll work up a repro for the assembly loading issue mentioned in the issue I linked to. Once I have it, where should I file it? I'm assuming corefx would be preferred over this repo?

**Comment by karelz** [comment]
@bording yeah, CoreFX might be better fit for now (as it is not tied to 4.7.1). We can move it later if it turns out to be wrong place. Please tag me and Alex on the issue. Thanks!

**Comment by misterspeedy** [comment]
We have had this issue running Xunit tests in a .NET 4.6.1 project.  No explicit references to 4.7 or standard or whatever.  We got a FileNotFoundException initializing XUnit.DiaSession.  I think this is confirmation of what @bording says above.  Problem started when Fall update installed, went away as soon as I rolled it back.

**Comment by firelizzard18** [comment]
I've run into a similar situation. One of my libraries, targeting .NET Standard 1.5, has a class with a method with a parameter of type `IEnumerable<(string Name, Type Type)>`. One of my applications, targeting .NET Framework 4.6.2, uses this method.

I was working on this application last month and had no issues. I updated to 1709 recently. Today, I started working on this application again and started getting a `MissingMethodException`:

```
System.MissingMethodException: Method not found: 'MyLibrary.MyClass.MyMethod(System.Type, System.Collections.Generic.IEnumerable`1<System.ValueTuple`2<System.String,System.Type>>)'.
```

The declaration is as follows:

```csharp
namespace MyLibrary
{
    public class MyClass
    {
        public static void MyMethod(Type type, IEnumerable<(string Name, Type Type)> parameters) {}
    }
}
```

Deleting the auto-generated binding redirect fixes the issue.
